Part1_InterruptControllersEvolution.html

```diff
@@ -104,79 +104,78 @@ <h3>PIC</h3>
+	<p>A special controller called LAPIC (Local APIC) was added for each processor, as well as the <a href="https://wiki.osdev.org/IOAPIC">I/O APIC</a> controller for routing interrupts from external devices. All of these controllers are combined in a common bus with the name APIC (note that modern systems use a standard system bus instead of a separate APIC bus for this task).<p/>
 
-	<p>When external interrupt arrives on the I/O APIC input, controller will send an interrupt message to LAPIC of one of the system CPUs. This way the I/O APIC controller help to ballance interrupt load between processors.<p/>
+	<p>When an external interrupt arrives on the I/O APIC input, the controller will send an interrupt message to the LAPIC of one of the system CPUs. In this way the I/O APIC controller helps balance interrupt load between processors.<p/>
 
-	<p>The first APIC chip was <a href="https://4donline.ihs.com/images/VipMasterIC/IC/INTL/INTLD047/INTLD047-2-1259.pdf?hkey=EF798316E3902B6ED9A73243A3159BB0">82489DX</a>, this was a separate chip with connected LAPIC and I/O APIC in itself. To make system with 2 processors it was needed 3 such chips (2 for LAPIC and 1 for I/O APIC). Later LAPIC functionality was directly included in processors, and an I/O APIC part was made to a separate 82093AA chip.<p/>
+	<p>The first APIC chip was the <a href="https://4donline.ihs.com/images/VipMasterIC/IC/INTL/INTLD047/INTLD047-2-1259.pdf?hkey=EF798316E3902B6ED9A73243A3159BB0">82489DX</a>, which was a separate chip that had a connected LAPIC and I/O APIC within itself. For a dual processor system three such chips were needed: two for LAPIC and one for I/O APIC. Later LAPIC functionality was directly included in processors, and the I/O APIC part was separated to the 82093AA chip.<p/>
 
-	<p>The I/O APIC <a href="https://pdos.csail.mit.edu/6.828/2016/readings/ia32/ioapic.pdf">82093AA</a> had 24 inputs, and APIC architecture could support up to 16 CPUs. For compatibility with older systems interrupts 0~15 were left for old ISA interrupts. And interrupts from PCI devices start to route on IRQ lines 16-23. With this you can forget about conflicts between ISA and PCI interrupts. Also cause of increased number of free interrupt lines it is became possible to increase number of PIRQx lines.<p/>
+	<p>The I/O APIC <a href="https://pdos.csail.mit.edu/6.828/2016/readings/ia32/ioapic.pdf">82093AA</a> had 24 inputs, and the APIC architecture could support up to 16 CPUs. For compatibility with older systems interrupts 0-15 were left for old ISA interrupts, and interrupts from PCI devices were routed on IRQ lines 16-23. This avoided all conflicts between ISA and PCI interrupts. With the increased number of free interrupt lines it also became possible to increase the number of PIRQx lines.<p/>

```

Comment:
"For compatibility with older systems interrupts 0-15 were left for old ISA interrupts, and interrupts from PCI devices were routed on IRQ lines 16-23"
Only the first part of the sentence is about compatibility: "For compatibility with older systems interrupts 0-15 were left for old ISA interrupts". The second one ("interrupts from PCI devices were routed on IRQ lines 16-23") is not about it.
I think this new sentence have lost it

Reply:
In that case we could rearrange it to say "Interrupts 0-15 were left for old ISA interrupts for compatibility with older systems, and interrupts from PCI devices were routed on IRQ lines 16-23."

Reply:
Can it be like so? Because in my opinion "started to route" is more correct, since this routing was optional. It wasn't a mandatory solution, just a good practice

```suggestion
	<p>The I/O APIC <a href="https://pdos.csail.mit.edu/6.828/2016/readings/ia32/ioapic.pdf">82093AA</a> had 24 inputs, and the APIC architecture could support up to 16 CPUs. Interrupts 0-15 were left for old ISA interrupts for compatibility with older systems, and interrupts from PCI devices started to route on IRQ lines 16-23. This avoided all conflicts between ISA and PCI interrupts. With the increased number of free interrupt lines it also became possible to increase the number of PIRQx lines.<p/>
```

Reply:
Alrighty, in that case we could just say "were optionally routed", which makes the optional part very clear

Reply:
This is once again about "good" and "bad" BIOS vendor. Good vendor would definitely do it and bad vendor might not.
What do you think about this rephrase?

```suggestion
	<p>The I/O APIC <a href="https://pdos.csail.mit.edu/6.828/2016/readings/ia32/ioapic.pdf">82093AA</a> had 24 inputs, and the APIC architecture could support up to 16 CPUs. Interrupts 0-15 were left for old ISA interrupts for compatibility with older systems, and interrupts 16-23 were meant for all the PCI devices. With this delimitation all conflicts between ISA and PCI interrupts could be easily avoided. With the increased number of free interrupt lines it also became possible to increase the number of PIRQx lines.<p/>
```